File: backend/app/services/academic_from_documents.py
"""Build academic fields from uploaded document extraction (no external mock APIs)."""
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Optional

from app.domain.document import Document
from app.domain.enums import DocType


def _num(value: Any) -> Optional[float]:
    if value is None or value == "":
        return None
    try:
        n = float(value)
    except (TypeError, ValueError):
        return None
    return n if n == n else None  # NaN guard


def _latest_doc(documents: list[Document], doc_type: DocType) -> Optional[Document]:
    matches = [d for d in documents if d.doc_type == doc_type]
    if not matches:
        return None
    return max(matches, key=lambda d: d.uploaded_at)


def _latest_parsed(documents: list[Document], doc_type: DocType) -> Optional[Document]:
    for doc in sorted(documents, key=lambda d: d.uploaded_at, reverse=True):
        if doc.doc_type != doc_type or not doc.extracted_data:
            continue
        if not isinstance(doc.extracted_data, dict):
            continue
        if any(k != "_missing" for k in doc.extracted_data):
            return doc
    return None


def yks_score_from_documents(documents: list[Document]) -> Optional[float]:
    yks = _latest_parsed(documents, DocType.YKS_RESULT) or _latest_doc(
        documents, DocType.YKS_RESULT
    )
    if not yks or not isinstance(yks.extracted_data, dict):
        return None
    data = yks.extracted_data
    if not data or all(k == "_missing" for k in data):
        return None
    return _num(
        data.get("placement_score")
        or data.get("score")
        or data.get("yks_score")
        or data.get("puan")
    )


def academic_record_from_documents(
    documents: list[Document],
) -> Optional[dict[str, Any]]:
    transcript = _latest_parsed(documents, DocType.TRANSCRIPT)
    yks_score = yks_score_from_documents(documents)

    if transcript is None and yks_score is None:
        return None

    t = transcript.extracted_data if transcript and isinstance(transcript.extracted_data, dict) else {}
    sources: list[str] = []

    record: dict[str, Any] = {
        "institution": None,
        "gpa_4": None,
        "gpa_100": None,
        "yks_score": None,
        "credits_completed": None,
        "fetched_at": datetime.now(timezone.utc).isoformat(),
        "source": None,
        "errors": None,
    }

    gpa = _num(t.get("gpa") or t.get("gpa_4"))
    if gpa is not None:
        record["gpa_4"] = gpa
        sources.append("TRANSCRIPT")
    if t.get("institution"):
        record["institution"] = str(t["institution"])
        if "TRANSCRIPT" not in sources:
            sources.append("TRANSCRIPT")
    credits = _num(t.get("completed_credits") or t.get("credits_completed"))
    if credits is not None:
        record["credits_completed"] = int(credits)

    if yks_score is not None:
        record["yks_score"] = yks_score
        sources.append("YKS")

    if not sources:
        return None

    record["source"] = f"{'+'.join(sources)} (parsed from documents)"
    return record

File: frontend/src/lib/academicFromDocuments.ts
import type { AcademicRecord, Document } from '../types/application'

function num(value: unknown): number | null {
  if (value == null || value === '') return null
  const n = Number(value)
  return Number.isFinite(n) ? n : null
}

function latestDoc(documents: Document[], docType: Document['doc_type']): Document | undefined {
  return documents
    .filter(d => d.doc_type === docType && d.extracted_data && Object.keys(d.extracted_data).length > 0)
    .sort((a, b) => new Date(b.uploaded_at).getTime() - new Date(a.uploaded_at).getTime())[0]
}

/** YKS score strictly from YKS_RESULT document extraction. */
export function yksScoreFromDocuments(documents: Document[]): number | null {
  const yks = latestDoc(documents, 'YKS_RESULT')
  if (!yks?.extracted_data) return null
  const data = yks.extracted_data
  return num(
    data.placement_score
    ?? data.score
    ?? data.yks_score
    ?? data.puan,
  )
}

/** Build academic fields from uploaded document extraction already stored on upload. */
export function academicRecordFromDocuments(documents: Document[]): AcademicRecord | null {
  const transcript = latestDoc(documents, 'TRANSCRIPT')
  const yksScore = yksScoreFromDocuments(documents)

  if (!transcript && yksScore == null) return null

  const t = transcript?.extracted_data ?? {}
  const sources: string[] = []

  const record: AcademicRecord = {
    institution: null,
    gpa_4: null,
    gpa_100: null,
    yks_score: null,
    credits_completed: null,
    fetched_at: new Date().toISOString(),
    source: null,
    errors: null,
  }

  const gpa = num(t.gpa ?? t.gpa_4)
  if (gpa != null) {
    record.gpa_4 = gpa
    sources.push('TRANSCRIPT')
  }
  if (t.institution) {
    record.institution = String(t.institution)
    if (!sources.includes('TRANSCRIPT')) sources.push('TRANSCRIPT')
  }
  const credits = num(t.completed_credits ?? t.credits_completed)
  if (credits != null) {
    record.credits_completed = credits
  }
  if (yksScore != null) {
    record.yks_score = yksScore
    sources.push('YKS')
  }

  if (sources.length === 0) return null
  record.source = `${sources.join('+')} (parsed from documents)`
  return record
}

/** Parsed document values override API/mock values when present. */
export function mergeAcademicRecord(
  fromApi: AcademicRecord | null,
  fromDocs: AcademicRecord | null,
  documents: Document[] = [],
): AcademicRecord | null {
  const yksFromPdf = yksScoreFromDocuments(documents)

  if (!fromApi && !fromDocs) return null

  const base = fromApi ?? fromDocs!
  const docFields = fromDocs ?? {
    institution: null,
    gpa_4: null,
    gpa_100: null,
    yks_score: null,
    credits_completed: null,
    fetched_at: null,
    source: null,
    errors: null,
  }

  const merged: AcademicRecord = {
    ...base,
    gpa_4: docFields.gpa_4 ?? fromApi?.gpa_4 ?? null,
    institution: docFields.institution ?? fromApi?.institution ?? null,
    credits_completed: docFields.credits_completed ?? fromApi?.credits_completed ?? null,
    yks_score: yksFromPdf ?? docFields.yks_score ?? fromApi?.yks_score ?? null,
    source: docFields.source ?? fromApi?.source ?? null,
  }

  const sourceParts: string[] = []
  if (merged.gpa_4 != null && latestDoc(documents, 'TRANSCRIPT')) sourceParts.push('TRANSCRIPT')
  if (merged.yks_score != null && yksFromPdf != null) sourceParts.push('YKS')
  if (sourceParts.length > 0) {
    merged.source = `${sourceParts.join('+')} (parsed from documents)`
  }

  return merged
}
